cpp/src/parquet/encoding.cc

```diff
@@ -354,6 +357,7 @@ class PlainEncoder<BooleanType> : public EncoderImpl, virtual public BooleanEnco
                                                       sink_.length(), n_valid);
```

Comment:
this line also seems problematic if this method is called multiple times with in a row with boolean arrays (not sure actual code does this though).

Reply:
https://github.com/apache/arrow/pull/36972#discussion_r1285158838 
You're right. This would not produce bug if PutArrow is not mixed with PutImpl, but will make Boolean leaves a larger space than expected

Reply:
You mentioned this below, using sink_.length() to store the actual values seems wrong.